crates/libs/windows/src/core/event.rs

```diff
@@ -0,0 +1,254 @@
+use super::*;
+use bindings::*;
+use std::sync::*;
+
+/// A type that you can use to declare and implement an event of a specified delegate type.
+///
+/// The implementation is thread-safe and designed to avoid contention between events being
+/// raised and delegates being added or removed.
+pub struct Event<T: Interface + Clone> {
+    swap: Mutex<()>,
```

Comment:
I understand the need for the `swap` and the `change` lock being different (to avoid contention), but why do you need the `swap` mutex instead of wrapping `delegates` array itself in a `RWLock`?

Reply:
Because both locks govern the same memory and their lifetimes are overlapping. The one handles changes and the other replacements. I don't think a `RwLock` can easily model this.